### PR TITLE
introduce fs-err for sake of better file errors

### PR DESCRIPTION
**Pull request by drahnr**
Ref mozilla/sccache#885

### CODE REVIEW

**Review by TriplEight — DISMISSED** [review]

**Review by TriplEight — APPROVED** [review]
LGTM

**Comment by drahnr** [comment]
Should be merged right after #31

**Comment by Xanewok** [comment]
Pushed https://github.com/paritytech/sccache/pull/19/commits/36b00f4ee2f3c9435c0f4840b345223e995051d6 and https://github.com/paritytech/sccache/pull/19/commits/e720b9a66e33699c892c469adb321cb9a35f2cbb to see if this fixes CI (can confirm locally)

File: src/dist/http.rs [hunk 1]
```diff
@@ -1454,43 +1456,35 @@ mod tests {
             }
         }
 
-        struct PersistOnPanic(Option<tempfile::NamedTempFile>);
```
Comment (Xanewok) [inline src/dist/http.rs:1457]: This seems to have been lost during rebase

File: src/test/utils.rs [hunk 2]
```diff
@@ -249,7 +248,7 @@ pub(crate) trait Waiter<R> {
 
 impl<T, O> Waiter<O> for T
 where
-    T: Future<Output = O>,
+    T: std::future::Future<Output = O>,
```
Comment (Xanewok) [inline src/test/utils.rs:251]: Lost in the rebase as well?

**Review by Xanewok — COMMENTED** [review]
I continue to have mixed feelings about this...

If this was mostly to replace `File::open(T: &Path)` and similar with wrapper functions that return a regular `File` but with an automatically created better messages then I'd feel a lot safer/better.
However, using `fs_err`:
1. seems to spread throughout our public API (because it introduces its own `File` wrapper)
2. its use already caught me off-guard and
3. it seems that this small rewrite potentially introduced a small scope(?)-related bug in `src/dist/cache.rs`, because we tried to work around the fact that `tempfile` works with regular `std::fs::Path`;
4. using the wrapper type and then converting to `std::fs::Path` boundary forces us to call `into_parts().0` (I glanced at the API and didn't notice a better conversion API, sorry if there is one) which seems cryptic and which we could solve already with "proper" error "handling" calls like `with_context(...)`

Maybe we could introduce our local util wrapper that automatically only wraps the filename in the error variant only for the `File::{open, create}` APIs? That should cover most of the uses already and seems like a lower impact change, instead.

File: src/lru_disk_cache/mod.rs [hunk 3]
```diff
@@ -279,7 +281,7 @@ impl LruDiskCache {
 
     /// Get an opened `File` for `key`, if one exists and can be opened. Updates the LRU state
     /// of the file if present. Avoid using this method if at all possible, prefer `.get`.
-    pub fn get_file<K: AsRef<OsStr>>(&mut self, key: K) -> Result<File> {
+    pub fn get_file<K: AsRef<OsStr>>(&mut self, key: K) -> Result<std::fs::File> {
```
[the remark this answers is not in the document]
Comment (drahnr) [reply src/lru_disk_cache/mod.rs:284]: I disagree. There is no consumer of this API other than ourselves. I'd rather re-export our `fs_err` as `sccache::fs` than going through the hassle.

File: src/compiler/compiler.rs [hunk 4]
```diff
@@ -1148,7 +1148,7 @@ mod test {
     use crate::mock_command::*;
     use crate::test::mock_storage::MockStorage;
     use crate::test::utils::*;
-    use std::fs::{self, File};
+    use crate::util::fs::{self, File};
```
Comment (drahnr) [inline src/compiler/compiler.rs:1151]: I'd be in favour of never using `File` but always stick to `fs::File` everywhere. There are almost no files with more than 2 occurences of `File`.

File: src/commands.rs [hunk 5]
```diff
@@ -665,7 +668,7 @@ pub fn run_command(cmd: Command) -> Result<i32> {
             let jobserver = unsafe { Client::new() };
             let creator = ProcessCommandCreator::new(&jobserver);
             let env: Vec<_> = env::vars_os().collect();
-            let out_file = File::create(out)?;
+            let out_file = fs::File::create(out).map(|x| x.into_parts().0)?;
```
[the remark this answers is not in the document]
Comment (Xanewok) [reply src/commands.rs]: Solved now that everything uses `fs_err`

File: src/test/utils.rs [hunk 6]
```diff
@@ -150,7 +150,7 @@ pub fn mk_bin_contents<F: FnOnce(File) -> io::Result<()>>(
     path: &str,
     fill_contents: F,
 ) -> io::Result<PathBuf> {
-    use std::os::unix::fs::OpenOptionsExt;
+    use fs::os::unix::fs::OpenOptionsExt;
```
Comment (drahnr) [inline src/test/utils.rs:153]: :+1:

File: src/compiler/compiler.rs (hunk 4 above)
Comment (Xanewok) [reply src/compiler/compiler.rs:1151]: This is more of a code style issue (arguments go both ways) and I'd rather avoid the churn of making changes for the sake of it. Let's keep the diff to a minimum and focus on the easily swapping out `std::fs` for better errors, which is the reason for this PR.

**Comment by Xanewok** [comment]
@drahnr are you fine with me merging this as-is?

File: src/compiler/compiler.rs (hunk 4 above)
Comment (drahnr) [reply src/compiler/compiler.rs:1151]: Fair point.

**Comment by drahnr** [comment]
Yes :+1: -> merged

**Pull request merged**